packages/binding-coap/src/coap-server.ts

```diff
@@ -337,387 +335,478 @@ export default class CoapServer implements ProtocolServer {
+            );
+            // TODO: Check if this has been fixed in the meantime
+            // node-coap does not support GET cancellation
+            this.sendErrorResponse(res, "5.01", "node-coap issue: no GET cancellation, send RST");
         }
+    }
 
-        // resource not found
-        res.code = "4.04";
-        res.end("Not Found");
+    /**
+     * Work-around to avoid duplicate requests in the case of observe
+     * registration (resend due to no response)
+     *
+     * (node-coap does not deduplicate when Observe is set)
+     *
+     * @param res The response that is being prepared for the observe registration request
+     */
+    private avoidDuplicatedObserveRegistration(res: OutgoingMessage) {
+        const packet = res._packet;
+        packet.code = "0.00";
+        packet.payload = Buffer.from("");
+        packet.reset = false;
+        packet.ack = true;
+        packet.token = Buffer.alloc(0);
+
+        res._send(res, packet);
+
+        res._packet.confirmable = res._request.confirmable;
+        res._packet.token = res._request.token;
     }
 
-    private isEmpty(obj: Record<string, unknown>) {
-        for (const key in obj) {
-            if (Object.prototype.hasOwnProperty.call(obj, key)) return false;
+    private getContentTypeFromRequest(req: IncomingMessage): string {
+        const contentType = req.headers["Content-Format"] as string;
+
+        if (contentType == null) {
+            warn(
+                `CoapServer on port ${this.getPort()} received no Content-Format from ${Helpers.toUriLiteral(
+                    req.rsinfo.address
+                )}:${req.rsinfo.port}`
+            );
         }
-        return true;
+
+        return contentType ?? ContentSerdes.DEFAULT;
+    }
+
+    private checkContentTypeSupportForInput(method: string, contentType: string) {
+        const methodsWithPayload: string[] = ["PUT", "POST", "FETCH", "iPATCH", "PATCH"];
+        const notAMethodWithPayload = !methodsWithPayload.includes(method);
+
+        return notAMethodWithPayload || ContentSerdes.get().isSupported(contentType);
+    }
+
+    private getThingDescriptionPayload() {
+        return Helpers.getAddresses().flatMap((address) =>
+            Array.from(this.things.keys()).map(
+                (thingKey) =>
+                    `${this.scheme}://${Helpers.toUriLiteral(address)}:${this.getPort()}/${encodeURIComponent(
+                        thingKey
+                    )}`
+            )
+        );
+    }
+
+    private parseUriSegments(requestUri: string) {
+        const segments = decodeURI(requestUri).split("/");
+
+        return {
+            thingKey: segments[1],
+            affordanceType: segments[2],
+            affordanceKey: segments[3],
+        };
+    }
+
+    private sendContentResponse(res: OutgoingMessage, payload: Buffer | string, contentType: string) {
+        res.setOption("Content-Format", contentType);
+        res.code = "2.05";
+        res.end(payload);
```

Comment:
VERY MINOR
In the following code there are some parts that always read like

```
        res.code = ...;
        res.end(payload);
```

It could make sense to have an overarching method that deals with all of them.. so that all point to it..

Reply:
Good point, I now turned the `sendErrorResponse` method into a generic `sendResponse` method :)